### ISSUE TITLE
Internal diff segfaults for large files with large number of changes

### ISSUE THREAD

**Issue by YodaEmbedding** [lewis6991/gitsigns.nvim#297]
**Describe the bug**
Internal diff segfaults for large files with large number of changes.

**To Reproduce**

init.vim:

```vim
let $PLUGIN_DIRECTORY = '~/.config/nvim/bundle'
set runtimepath^=$PLUGIN_DIRECTORY/plenary.nvim
set runtimepath^=$PLUGIN_DIRECTORY/gitsigns.nvim

lua << EOF
require('gitsigns').setup {
  debug_mode = true,
  use_internal_diff = true,
}
EOF
```

```bash
mkdir tmp
cd tmp
seq 1 100000 > file
git init
git add file
git commit -m "init"
seq 1 10000 > file
sed -i 's/0/00000000000000000000/g' file
sed -i 's/0/00000000000000000000/g' file
sed -i 's/0/00000000000000000000/g' file
vim file
```

NOTE: You may need to modify the `seq` count or `sed` calls depending on your system's resources. e.g. On systems with more RAM (?), try adding more `0`s to the `sed` calls.

**Observed output**
SEGFAULT.

**Additional context**
System: Arch Linux
Neovim version: 0.5

**Comment by lewis6991** [comment]
Thanks for reporting.

Unfortunately I wasn't able to reproduce and this even with increasing the `seq` count by x100 and increasing the `sed` count too. This caused nvim to take around a minute to open but no segfault once it finallly did. Though interestingly I did get this message:

```
Error executing vim.schedule lua callback: not enough memory
Error executing luv callback:
...ck/packer/start/plenary.nvim/lua/plenary/async/async.lua:14: The coroutine failed with this message: not
 enough memory
```

Either way, this is either an issue with [diff_ffi.tl](https://github.com/lewis6991/gitsigns.nvim/blob/main/teal/gitsigns/diff_ffi.tl) or neovims xdiff library.

Are you able to provide a backtrace or a coredump? Would be good to know where the segfault is happening.

Also you may want to try #279, you will need to build nvim from source with the linked PR.

**Comment by YodaEmbedding** [comment]
Sorry, my reproduction case was missing a 3rd line of `sed` -- fixed above.

Funnily, when I compile neovim from source on `v0.5.0` tag, I get the luv error you mentioned but no segfault. The latest 0.6 seems to be completely fine for the given test case.

---

Backtrace for system-wide installation of 0.5:

<details>
<summary>bt</summary>

```
(gdb) bt
#0  0x00007ffff7c6b380 in __memchr_avx2 () from /usr/lib/libc.so.6
#1  0x00005555557fe83c in xdl_guess_lines (mf=mf@entry=0x41718400, sample=sample@entry=256) at /mnt/data/ProgData/.cache/paru/clone/neovim-git/src/neovim-git/src/nvim/xdiff/xutils.c:132
#2  0x0000555555801942 in xdl_prepare_env (mf1=0x41718400, mf2=0x41718420, xpp=xpp@entry=0x41718440, xe=xe@entry=0x7fffffffcef0)
    at /mnt/data/ProgData/.cache/paru/clone/neovim-git/src/neovim-git/src/nvim/xdiff/xprepare.c:279
#3  0x0000555555801b58 in xdl_do_diff (mf1=<optimized out>, mf2=<optimized out>, xpp=xpp@entry=0x41718440, xe=xe@entry=0x7fffffffcef0)
    at /mnt/data/ProgData/.cache/paru/clone/neovim-git/src/neovim-git/src/nvim/xdiff/xdiffi.c:320
#4  0x0000555555801d04 in xdl_diff (mf1=<optimized out>, mf2=<optimized out>, xpp=0x41718440, xecfg=0x42ac8b78, ecb=0x41718468)
    at /mnt/data/ProgData/.cache/paru/clone/neovim-git/src/neovim-git/src/nvim/xdiff/xdiffi.c:1017
#5  0x00007ffff7ce5abb in ?? () from /usr/lib/libluajit-5.1.so.2
#6  0x00007ffff7d1b48d in ?? () from /usr/lib/libluajit-5.1.so.2
#7  0x00007ffff7d1bd40 in ?? () from /usr/lib/libluajit-5.1.so.2
#8  0x00007ffff7ce3936 in ?? () from /usr/lib/libluajit-5.1.so.2
#9  0x00007ffff7ce4453 in ?? () from /usr/lib/libluajit-5.1.so.2
#10 0x00007ffff7d29790 in lua_pcall () from /usr/lib/libluajit-5.1.so.2
#11 0x00005555556d5e1d in nlua_luv_cfpcall (lstate=0x40000378, nargs=0, nresult=0, flags=0) at /mnt/data/ProgData/.cache/paru/clone/neovim-git/src/neovim-git/src/nvim/lua/executor.c:225
#12 0x00007ffff7f38841 in ?? () from /usr/lib/libuv.so.1
#13 0x00007ffff7f318a9 in uv_run () from /usr/lib/libuv.so.1
#14 0x00005555556680bf in loop_poll_events (loop=0x5555558f8920 <main_loop>, ms=3953) at /mnt/data/ProgData/.cache/paru/clone/neovim-git/src/neovim-git/src/nvim/event/loop.c:62
#15 0x000055555573d429 in inbuf_poll (ms=4000, events=events@entry=0x55555590e400) at /mnt/data/ProgData/.cache/paru/clone/neovim-git/src/neovim-git/src/nvim/os/input.c:415
#16 0x000055555573da18 in os_inchar (buf=0x0, maxlen=0, ms=-1, tb_change_cnt=0, events=0x55555590e400)
    at /mnt/data/ProgData/.cache/paru/clone/neovim-git/src/neovim-git/src/nvim/os/input.c:118
#17 0x00005555557b07f0 in state_enter (s=0x7fffffffd580) at /mnt/data/ProgData/.cache/paru/clone/neovim-git/src/neovim-git/src/nvim/state.c:54
#18 0x000055555570d751 in normal_enter (cmdwin=<optimized out>, noexmode=<optimized out>) at /mnt/data/ProgData/.cache/paru/clone/neovim-git/src/neovim-git/src/nvim/normal.c:464
#19 0x00005555556e123c in main (argc=-10512, argv=<optimized out>) at /mnt/data/ProgData/.cache/paru/clone/neovim-git/src/neovim-git/src/nvim/main.c:562
```

</details>

<details>
<summary>bt full</summary>

```
(gdb) bt full
#0  0x00007ffff7c6b380 in __memchr_avx2 () from /usr/lib/libc.so.6
No symbol table info available.
#1  0x00005555557fe83c in xdl_guess_lines (mf=mf@entry=0x41718400, sample=sample@entry=256) at /mnt/data/ProgData/.cache/paru/clone/neovim-git/src/neovim-git/src/nvim/xdiff/xutils.c:132
        nl = 1
        size = 588895
        tsize = 0
        data = 0x42041020 <error: Cannot access memory at address 0x42041020>
        cur = <optimized out>
        top = 0x420d0c7f ""
#2  0x0000555555801942 in xdl_prepare_env (mf1=0x41718400, mf2=0x41718420, xpp=xpp@entry=0x41718440, xe=xe@entry=0x7fffffffcef0)
    at /mnt/data/ProgData/.cache/paru/clone/neovim-git/src/neovim-git/src/nvim/xdiff/xprepare.c:279
        enl1 = <optimized out>
        enl2 = <optimized out>
        sample = 256
        cf = {hbits = 0, hsize = 0, rchash = 0x0, ncha = {head = 0x0, tail = 0x0, isize = 0, nsize = 0, ancur = 0x0, sncur = 0x0, scurr = 0}, rcrecs = 0x0, alloc = 0, count = 0, flags = 0}
#3  0x0000555555801b58 in xdl_do_diff (mf1=<optimized out>, mf2=<optimized out>, xpp=xpp@entry=0x41718440, xe=xe@entry=0x7fffffffcef0)
    at /mnt/data/ProgData/.cache/paru/clone/neovim-git/src/neovim-git/src/nvim/xdiff/xdiffi.c:320
        ndiags = <optimized out>
        kvd = <optimized out>
        kvdf = <optimized out>
        kvdb = <optimized out>
        xenv = {mxcost = 3472328296227680304, snake_cnt = 3472328296227680304, heur_min = 3472328296227680304}
        dd1 = {nrec = 1097954888, ha = 0x7ffff7c47574 <call_dl_lookup+52>, rindex = 0x7, rchg = 0x0}
        dd2 = {nrec = 140737488343040, ha = 0x0, rindex = 0x7ffff7d17436, rchg = 0x0}
#4  0x0000555555801d04 in xdl_diff (mf1=<optimized out>, mf2=<optimized out>, xpp=0x41718440, xecfg=0x42ac8b78, ecb=0x41718468)
    at /mnt/data/ProgData/.cache/paru/clone/neovim-git/src/neovim-git/src/nvim/xdiff/xdiffi.c:1017
        xscr = 0xfc85e1bc68a0aaba
        xe = {xdf1 = {rcha = {head = 0xffffffffffffffd8, tail = 0x7ffff7d17436, isize = 0, nsize = 1097954888, ancur = 0xfc85e1bc55e0aaba, sncur = 0xfc85f1cb3102aaba, scurr = 0},
            nrec = 140737350936181, hbits = 808464432, rhash = 0x7fffffffd1c0, dstart = 140737488343488, dend = 140737350905279, recs = 0x0, rchg = 0x0, rindex = 0x400b39b8,
            nreff = 1097954888, ha = 0x4171b058}, xdf2 = {rcha = {head = 0x7ffff7d17fb8, tail = 0xfffffffffffffeb8, isize = 140733193389312, nsize = 1097958504, ancur = 0x7fffffffd230,
              sncur = 0x41715568, scurr = -6068330269344955648}, nrec = 140737352979456, hbits = 1280, rhash = 0x41717648, dstart = 1097970040, dend = 140737488343408, recs = 0x400b3ec8,
            rchg = 0x400b3ec8 "`\204qA\365\377\377\377p\235qAr", rindex = 0x7ffff7d196b1, nreff = 0, ha = 0x55555555eb20}}
        ef = 0x5555557fe4e4 <xdl_call_hunk_func>
#5  0x00007ffff7ce5abb in ?? () from /usr/lib/libluajit-5.1.so.2
No symbol table info available.
#6  0x00007ffff7d1b48d in ?? () from /usr/lib/libluajit-5.1.so.2
No symbol table info available.
```

</details>

**Comment by lewis6991** [comment]
Good work.

Doesn't seem like the issue is due to gitsigns or even neovim.

The segfault originates deep within the xdiff lib. It could be we are just pushing it too far?

Can you reproduce the issue using #301 with the neovim PR that adds xdiff bindings? That'll rule out FFI causing the problem.

**Comment by YodaEmbedding** [comment]
Building #301 and https://github.com/neovim/neovim/pull/14536 manually from source does not segfault but gives:

```
Error executing luv callback:
...fig/nvim/bundle/plenary.nvim/lua/plenary/async/async.lua:14: The coroutine failed with this message: ...nfig/nvim/bundle/gitsigns.nvim/lua/gitsigns/diff_int.lua:19: bad argument #1 to 'run
_diff_xdl' (expected string)
```

I also tried another system-wide install with these PRs applied, and this time, it gives the same error as above.

```diff
diff --git a/PKGBUILD b/PKGBUILD
index a9cbe4b..e13625f 100644
--- a/PKGBUILD
+++ b/PKGBUILD
@@ -30,6 +30,13 @@ pkgver() {
   git describe --long | sed 's/^v//;s/\([^-]*-g\)/r\1/;s/-/./g'
 }
 
+prepare() {
+  cd "${pkgname}"
+  git remote remove origin
+  git remote add origin git@github.com:neovim/neovim.git
+  gh pr checkout 14536
+}
+
 build() {
   cmake -S"${pkgname}" -Bbuild \
         -DCMAKE_BUILD_TYPE=RelWithDebInfo \
```

**Comment by lewis6991** [comment]
> Building #301 and neovim/neovim#14536 manually from source does not segfault but gives:

Woops, #301 had a bug in it, just updated, try again.

**Comment by YodaEmbedding** [comment]
After applying the latest patches for the PRs mentioned, no issues with both manual source build and system-wide PKGBUILD build on latest neovim master.

**Comment by lewis6991** [comment]
I'm a C novice, but it seems the issue is probably due to some internal limits of LuaJIT's FFI not being able to allocate memory.

Will close this when the mentioned PR's get closed.

**Comment by lewis6991** [comment]
Should be fixed with #301
